kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1180 @@
+            {
+                ModelPart::ConditionsContainerType::iterator it = cond_begin + k;
+
+                //detect if the element is active or not. If the user did not make any choice the element
+                //is active by default
+                bool condition_is_active = true;
+                if ((it)->IsDefined(ACTIVE))
+                    condition_is_active = (it)->Is(ACTIVE);
+
+                if (condition_is_active)
+                {
+                    //calculate elemental contribution
+                    pScheme->Condition_CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
+                    ApplyConstraints<Condition>(rModelPart, *it, LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
+
+                    //assemble the elemental contribution
+#ifdef USE_LOCKS_IN_ASSEMBLY
+                    this->Assemble(A, b, LHS_Contribution, RHS_Contribution, EquationId, BaseType::mlock_array);
+#else
+                    this->Assemble(A, b, LHS_Contribution, RHS_Contribution, EquationId);
+#endif
+
+                    // clean local elemental memory
+                    pScheme->CleanMemory(*(it.base()));
+                }
+            }
+        }
+
+        const double stop_build = OpenMPUtils::GetCurrentTime();
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverWithConstraints", (this->GetEchoLevel() >= 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Build time: " << stop_build - start_build << std::endl;
+
+        //for (int i = 0; i < A_size; i++)
+        //    omp_destroy_lock(&lock_array[i]);
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverWithConstraints", (this->GetEchoLevel() > 2 && rModelPart.GetCommunicator().MyPID() == 0)) << "Finished parallel building" << std::endl;
+
+        KRATOS_CATCH("")
+    }
+
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+  private:
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    // This is the set of condenced global constraints.
+    GlobalMasterSlaveRelationContainerType mGlobalMasterSlaveRelations; //This can be changed to more efficient implementation later on.
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief   this method condences the MasterSlaveConstraints which are added on the rModelPart
+     *          into objects of AuxilaryGlobalMasterSlaveRelation. One unique object for each unique slave.
+     *          these will be used in the ApplyConstraints functions late on.
+     *          matrix and the right hand side
+     * @param   rModelPart The model part of the problem to solve
+     */
+    void FormulateGlobalMasterSlaveRelations(ModelPart& rModelPart)
```

Comment:
By "Assemble" I mean to convey exactly the FEM type of assemble. Here in this AssembleSlaves()
two MasterSlaveRelation objects 

s1 = m1*w1 + m3*w3
and 
s1 = m4*w4 + m2*w2 + m3*w33 

will be assembled (condensed) into 

s1 = m1*w1 +  m2*w2 + m3*(w3+w33) + m4*w4

From FormulateGlobalMasterSlaveRelations to SetupGlobalMasterSlaveRelations ... I have no objection.

Reply:
I see, still as a developer who is not familiar with your code, I think it would be helpful to reflect that `AssembleSlaves` is acting on the ConstraintRelation directly in the name. Would it be reasonable to call it `AssembleConstraintRelation`, i.e., `SetUpSystem()->SetUpConstraintRelations()->AssembleConstraintRelation()`?